--- a/configure
+++ b/configure
@@ -11427,10 +11427,10 @@ if test "$USE_INCLUDED_LIBINTL" = "yes" ; then
 fi

 if test $cross_compiling = no; then
-   BUILD_CFLAGS="$CFLAGS $CPPFLAGS"
+   BUILD_CFLAGS="$CFLAGS $CPPFLAGS $INCLUDES -DHAVE_CONFIG_H"
    BUILD_LDFLAGS="$LDFLAGS"
 else
-   BUILD_CFLAGS=
+   BUILD_CFLAGS="$INCLUDES"
    BUILD_LDFLAGS=
 fi

diff --git a/debugfs/debugfs.c b/debugfs/debugfs.c
index 80662b1..ac37275 100644
--- a/debugfs/debugfs.c
+++ b/debugfs/debugfs.c
@@ -26,6 +26,9 @@ extern char *optarg;
 #include <fcntl.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#ifdef HAVE_SYS_SYSMACROS_H
+#include <sys/sysmacros.h>
+#endif

 #include "debugfs.h"
 #include "uuid/uuid.h"
diff --git a/lib/blkid/devname.c b/lib/blkid/devname.c
index b151354..3c4173b 100644
--- a/lib/blkid/devname.c
+++ b/lib/blkid/devname.c
@@ -35,6 +35,9 @@
 #if HAVE_SYS_MKDEV_H
 #include <sys/mkdev.h>
 #endif
+#ifdef HAVE_SYS_SYSMACROS_H
+#include <sys/sysmacros.h>
+#endif
 #include <time.h>

 #include "blkidP.h"
diff --git a/lib/blkid/devno.c b/lib/blkid/devno.c
index c9f5c92..91c9207 100644
--- a/lib/blkid/devno.c
+++ b/lib/blkid/devno.c
@@ -30,6 +30,9 @@
 #if HAVE_SYS_MKDEV_H
 #include <sys/mkdev.h>
 #endif
+#ifdef HAVE_SYS_SYSMACROS_H
+#include <sys/sysmacros.h>
+#endif

 #include "blkidP.h"

diff --git a/lib/ext2fs/ismounted.c b/lib/ext2fs/ismounted.c
index db80c46..247b9ff 100644
--- a/lib/ext2fs/ismounted.c
+++ b/lib/ext2fs/ismounted.c
@@ -37,6 +37,9 @@
 #endif /* HAVE_GETMNTINFO */
 #include <string.h>
 #include <sys/stat.h>
+#ifdef HAVE_SYS_SYSMACROS_H
+#include <sys/sysmacros.h>
+#endif

 #include "ext2_fs.h"
 #include "ext2fs.h"